core/node/vm_runner/src/process.rs

```diff
@@ -0,0 +1,153 @@
+/// already been processed by state keeper but still require some extra handling as regulated by
+/// [`OutputHandlerFactory`].
+///
+/// It's responsible for taking unprocessed data from the [`VmRunnerIo`], feeding it into
+/// [`BatchExecutor`] and calling [`OutputHandlerFactory`] on the result of the execution (batch
+/// execution state in the [`UpdatesManager`]).
+///
+/// You can think of VM runner as a concurrent processor of a continuous stream of newly committed
+/// batches/blocks.
+#[derive(Debug)]
+pub struct VmRunner {
+    pool: ConnectionPool<Core>,
+    io: Box<dyn VmRunnerIo>,
+    storage: Arc<dyn StorageLoader>,
+    output_handler_factory: Box<dyn OutputHandlerFactory>,
+    batch_processor: Box<dyn BatchExecutor>,
+}
+
+impl VmRunner {
+    /// Initializes VM runner with its constituents. In order to make VM runner concurrent each
+    /// parameter here needs to support concurrent execution mode. See
+    /// [`ConcurrentOutputHandlerFactory`], [`VmRunnerStorage`].
+    ///
+    /// Caller is expected to provide a component-specific implementation of [`VmRunnerIo`] and
+    /// an underlying implementation of [`OutputHandlerFactory`].
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        io: Box<dyn VmRunnerIo>,
+        storage: Arc<dyn StorageLoader>,
+        output_handler_factory: Box<dyn OutputHandlerFactory>,
+        batch_processor: Box<dyn BatchExecutor>,
+    ) -> Self {
+        Self {
+            pool,
+            io,
+            storage,
+            output_handler_factory,
+            batch_processor,
+        }
+    }
+
+    /// Consumes VM runner to execute a loop that continuously pulls data from [`VmRunnerIo`] and
+    /// processes it.
+    pub async fn run(mut self, stop_receiver: &watch::Receiver<bool>) -> anyhow::Result<()> {
+        const SLEEP_INTERVAL: Duration = Duration::from_millis(50);
+
+        let mut next_batch = self
+            .io
+            .latest_processed_batch(&mut self.pool.connection().await?)
+            .await?
+            + 1;
+        loop {
+            let last_ready_batch = self
+                .io
+                .last_ready_to_be_loaded_batch(&mut self.pool.connection().await?)
+                .await?;
+            if next_batch > last_ready_batch {
+                // Next batch is not ready to be processed yet
+                tokio::time::sleep(SLEEP_INTERVAL).await;
+                continue;
+            }
+            let Some(batch_data) = self.storage.load_batch(next_batch).await? else {
+                // Next batch has not been loaded yet
+                tokio::time::sleep(SLEEP_INTERVAL).await;
+                continue;
+            };
+            let mut updates_manager =
+                UpdatesManager::new(&batch_data.l1_batch_env, &batch_data.system_env);
+            let Some(handler) = self
+                .batch_processor
+                .init_batch(
```

Comment:
Hmm, from what I see, errors in `init_batch` can only happen in the command receiver task. Or do you mean that we should represent interruption as a `Result` rather than `Option`?

Reply:
Ah, sorry; I've looked at `init_batch()` code not noticing that smelling `expect`s are performed in a separate task. It would make sense to propagate errors in that task though, but it definitely looks like something worthy a separate PR.